kaiax/gov/impl/init.go

```diff
@@ -0,0 +1,74 @@
+package impl
+
+import (
+	"math/big"
+
+	"github.com/kaiachain/kaia/blockchain/state"
+	"github.com/kaiachain/kaia/blockchain/types"
+	"github.com/kaiachain/kaia/common"
+	"github.com/kaiachain/kaia/kaiax/gov"
+	"github.com/kaiachain/kaia/kaiax/gov/contractgov"
+	"github.com/kaiachain/kaia/kaiax/gov/headergov"
+	"github.com/kaiachain/kaia/log"
+	"github.com/kaiachain/kaia/params"
+)
+
+var (
+	_ gov.GovModule = (*GovModule)(nil)
+
+	logger = log.NewModuleLogger(log.KaiaXGov)
+)
+
+//go:generate mockgen -destination=kaiax/gov/impl/mock/blockchain_mock.go github.com/kaiachain/kaia/kaiax/gov/impl BlockChain
+type BlockChain interface {
+	CurrentBlock() *types.Block
+	Config() *params.ChainConfig
+	GetHeaderByNumber(val uint64) *types.Header
+	StateAt(root common.Hash) (*state.StateDB, error)
+	GetReceiptsByBlockHash(blockHash common.Hash) types.Receipts
+	GetBlock(hash common.Hash, number uint64) *types.Block
+}
+
+type GovModule struct {
+	hgm   headergov.HeaderGovModule
+	cgm   contractgov.ContractGovModule
+	chain BlockChain
+}
+
+type InitOpts struct {
+	Hgm   headergov.HeaderGovModule
+	Cgm   contractgov.ContractGovModule
+	Chain BlockChain
+}
+
+func NewGovModule() *GovModule {
+	return &GovModule{}
+}
+
+func (m *GovModule) Init(opts *InitOpts) error {
+	if opts == nil {
+		return gov.ErrInitNil
+	}
+
+	m.hgm = opts.Hgm
+	m.cgm = opts.Cgm
+	m.chain = opts.Chain
+
+	if m.hgm == nil || m.cgm == nil || m.chain == nil || m.chain.Config() == nil {
+		return gov.ErrInitNil
+	}
+	return nil
+}
+
+func (m *GovModule) Start() error {
```

Comment:
Good question @hyunsooda. I intentionally designed Init and Start to be invoked at different timing. You might not see their usage at this stage, but they do serve different purposes.

- `Constructor`: to declare module variables so Init() functions can interlink modules.
- `Init`: to link modules (kaiax/*) and components (BlockChain, ChainDB, etc)
  Having both Constructor and Init enables circular dependencies. e.g.
	```go
	mValset = NewValsetModule()
	mGov = NewGovModule()
	
	mValset.Init(InitOpts{... mGov ...})
	mGov.Init(InitOpts{... mValset ...})
	```
    This two-step initialization is inspired by CosmosSDK's initialization code (https://github.com/cosmos/cosmos-sdk/blob/v0.50.10/simapp/app.go#L282-L425)
- `Start`: Start any background thread ___after___ Init is complete. There are exactly two call sites.
  - At `CN.Start()` after other components are started (i.e. can receive internal requests such as Subscribe) -- because of the possibility of circular dependencies, you must wait until all modules are Init'ed, and then start your tasks.
  - At `CNAPIBackend.SetHead()` where modules are restarted after a chain rewind -- this one cannot be merged with Init()
- `Stop`: Stop any background thread. Similar to Start().

Yes, Start and Stop currently does nothing. But there will be some cases where a module maintains a long-running thread (e.g. SupplyManager which can be ported as kaiax/supply, ChainDataFetcher which can be ported as kaiax/cdf)